crates/ide/src/ide/assists/inline.rs

```diff
@@ -0,0 +1,89 @@
+use super::{AssistKind, AssistsCtx};
+use crate::def::{AstPtr, ResolveResult};
+use crate::TextEdit;
+use smol_str::ToSmolStr;
+use syntax::ast::AstNode;
+use syntax::{ast, match_ast};
+
+pub(super) fn inline(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let file_id = ctx.frange.file_id;
+    let parse = ctx.db.parse(file_id);
+
+    let covering_node = ctx.covering_node::<ast::Ref>()?;
```

Comment:
Let's do something like rust-analyzer:
1. On a reference identifier, the action replaces this reference with the definition. This is what is implemented by this PR currently.
2. On a definition identifier (of let-in), the action replaces all references and delete the definition. For simplicity, we can only support single-definition for now (excluding `let a.b = 1; a.b = 2; in ..` case).

Reply:
> Beside that, I can't seem to find the documentation for `SyntaxNodePtr::to_node(&root: SyntaxNode)` function. What role does `root` play here?

`SyntaxNodePtr` is basically a typed `TextRange` that only stores the start and end location. These locations need to be resolved in a root node, which is almost always the whole `SourceFile`. There is a shortcut `ctx.ast.syntax()` to easily get it. It is the same value as `db.parse(current_file_id)`.

> the deleted let-in will eat into the follow expressions

You can debug it by `dbg!(&rewrites)` and manually check the rewrite positions. See also the comment on the location use of `LetIn`.